### PR TITLE
KeepRepo -> KeepToLibrary migration

### PR DESCRIPTION
**Pull request by ryanpbrewster**
KeepRepo shouldn't need to know about libraries anymore.

@leogrim @andrewconner 

I'll go through and tag a few of the things I want feedback on.

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 1]
```diff
@@ -499,7 +497,7 @@ class KeepRepoImpl @Inject() (
 
   def librariesWithMostKeepsSince(count: Int, since: DateTime)(implicit session: RSession): Seq[(Id[Library], Int)] = {
     import com.keepit.common.db.slick.StaticQueryFixed.interpolation
-    sql"""select b.library_id, count(*) as cnt from bookmark b, library l where l.id = b.library_id and l.state='active' and l.visibility='published' and b.kept_at > $since group by b.library_id order by count(*) desc limit $count""".as[(Id[Library], Int)].list
+    sql"""select b.library_id, count(*) as cnt from bookmark b, library l where l.id = b.library_id and l.state='active' and l.visibility='published' and b.kept_at > $since group by b.library_id order by count(*) desc, b.library_id asc limit $count""".as[(Id[Library], Int)].list
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:500]: This method was non-deterministic before. it was undefined how to break ties if two libraries had equal numbers of new keeps. I added a tie-breaker strictly so that my backwards compatibility tests would pass.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 2]
```diff
@@ -528,14 +526,14 @@ class KeepRepoImpl @Inject() (
...
-        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at > ${after.keptAt} ORDER BY kept_at DESC LIMIT $limit;""".as[Keep].list
+        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at > ${after.keptAt} ORDER BY kept_at DESC, id DESC LIMIT $limit;""".as[Keep].list
       case (None, None) =>
-        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId ORDER BY kept_at DESC LIMIT $limit;""".as[Keep].list
+        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId ORDER BY kept_at DESC, id DESC LIMIT $limit;""".as[Keep].list
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:536]: @andrewconner I added a tie-breaker here.

I'll tag you in the version of this that I implemented in `KeepToLibraryRepo`. As we talked about before lunch, I think it's possible to break ties in a consistent manner.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 3]
```diff
@@ -126,4 +145,86 @@ class KeepToLibraryRepoImpl @Inject() (
...
+  /**
+   * ***************************************************
+   * For backwards compatibility with KeepRepo methods *
+   * ***************************************************
+   */
+
+  def getByLibraryIdAndExcludingVisibility(libId: Id[Library], excludeVisibility: Option[LibraryVisibility], limit: Limit)(implicit session: RSession): Seq[KeepToLibrary] = {
+    val q = { for (ktl <- rows if ktl.libraryId === libId && ktl.visibility =!= excludeVisibility.orNull) yield ktl }.take(limit.value)
+    q.list
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:158]: This method is used strictly for checking that Keeps are consistent with their Library.

I would _love_ to simply have a DB constraint that ensures that `KTL`s are consistent with the library they're attached to. The reason we don't do this now is (presumably) because it takes time to propagate a library change over to all the keeps it contains, so we can't have that constraint. Is there some clever DB thing we can do to get around this?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 4]
```diff
@@ -126,4 +145,86 @@ class KeepToLibraryRepoImpl @Inject() (
...
+    val recentLibraryKeeps = (beforeIdOpt.map(get), afterIdOpt.map(get)) match {
+      case (None, None) =>
+        allLibraryKeeps
+      case (Some(upper), _) =>
+        allLibraryKeeps.filter(ktl => ktl.addedAt < upper.addedAt || (ktl.addedAt === upper.addedAt && ktl.keepId < upper.keepId))
+      case (None, Some(lower)) =>
+        allLibraryKeeps.filter(ktl => ktl.addedAt > lower.addedAt || (ktl.addedAt === lower.addedAt && ktl.keepId > lower.keepId))
+    }
+    recentLibraryKeeps.sortBy(ktl => (ktl.addedAt desc, ktl.keepId desc)).take(limit.value).list
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:194]: @andrewconner this is the version I wrote. You have to provide it with a set of libraries, so we'd need to grab the user's followed libraries ahead of time. I think that should be really fast, though, especially if we can cache it.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 5]
```diff
@@ -126,4 +145,86 @@ class KeepToLibraryRepoImpl @Inject() (
...
+    } else {
+      val idset = libIds.map { _.id }.mkString("(", ",", ")")
+      val q = sql"""select ktl.library_id, max(bm.seq) 
+                    from keep_to_library ktl, bookmark bm
+                    where bm.id = ktl.keep_id and ktl.library_id in #${idset}
+                    group by ktl.library_id"""
+
+      q.as[(Long, Long)].list.map { case (libId, seq) => Id[Library](libId) -> SequenceNumber[Keep](seq) }.toMap
+    }
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:220]: @leogrim, I'm not very familiar with how sequence numbers are used. Is this something we want to maintain going forward?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 6]
```diff
@@ -528,14 +526,14 @@ class KeepRepoImpl @Inject() (
 
     (beforeIdOpt.flatMap(getOpt), afterIdOpt.flatMap(getOpt)) match {
       case (Some(before), _) =>
-        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at <= ${before.keptAt} AND id < ${before.id.get} ORDER BY kept_at DESC LIMIT $limit;""".as[Keep].list
+        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at <= ${before.keptAt} AND id < ${before.id.get} ORDER BY kept_at DESC, id DESC LIMIT $limit;""".as[Keep].list
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:529]: Oo, I think the original was broken, too. `before.id` may be higher or lower than the next keep.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:536]: Yeah this one is not a problem, since it's the first set (no before, no after).

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 7]
```diff
@@ -528,14 +526,14 @@ class KeepRepoImpl @Inject() (
...
       case (None, Some(after)) =>
         // This case is not strictly correct. It's not possible to call after a keep, and get other keeps kept in the same ms.
         // Fortunately, ending in this state where you have a keep id and need ones that happened after (and they happened in the same ms)
         // is nearly impossible. We can't use IDs as tie breakers because old IDs may get updated kept_at fields.
-        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at > ${after.keptAt} ORDER BY kept_at DESC LIMIT $limit;""".as[Keep].list
+        sql"""SELECT #$bookmarkColumnOrder FROM bookmark bm WHERE library_id IN (SELECT library_id FROM library_membership WHERE user_id=$userId AND state='active') AND state='active' AND user_id!=$userId AND kept_at > ${after.keptAt} ORDER BY kept_at DESC, id DESC LIMIT $limit;""".as[Keep].list
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:534]: I see this one doesn't have the `id` problem (that the above has), and is a weird edge case anyways, so I think this one's fine.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 5 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:220]: I'll take another look tomorrow morning (it's almost 2am) but this looks weird, when do we need this?
To drive an integrity plugin maybe?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 4 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:194]: Yep, that makes sense, we may also want to grab the user's connections to get keeps from their `Main` discoverable libraries.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:194]: I think it's right. A test that has a lot of these edge-case keeps (with out-of-order `id`s, equal `keep_at`s, etc) and just verifies that they all come out is the gold here.

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:194]: Agreed, I'll make an abusive test.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 5 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:220]: Yeah, take a look at the corresponding `KeepRepo` method. That's used in `SuggestedSearchTermScheduler.librariesNeedUpdate`.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 6 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:529]: Oy, we should fix that. I'm not sure where `latestKeep` is used but it probably isn't doing the right thing. ((Although _most_ Keeps are probably in order.))

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala (hunk 5 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:220]: I see, seems reasonable.

File: kifi-backend/modules/shoebox/test/com/keepit/model/KeepToLibraryRepoTest.scala [hunk 8]
```diff
@@ -56,5 +75,223 @@ class KeepToLibraryRepoTest extends Specification with ShoeboxTestInjector {
...
+            val libIds = libs.map(_.id.get).toSet
+            val expected = inject[KeepRepo].getMaxKeepSeqNumForLibraries(libIds)
+            val actual = inject[KeepToLibraryRepo].getMaxKeepSeqNumForLibraries(libIds)
+
+            expected === actual
+          }
+          1 === 1
+        }
+      }
+    }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/test/com/keepit/model/KeepToLibraryRepoTest.scala:295]: So many tests, it's Christmas already =)
